Implement DropGuard to ensure future Completion

leodziki opened the pull request:
# Description

To manage state transitions safely across `await` boundaries, we introduced a `DropGuard` wrapper. This ensures that even if a future is dropped before completion, it will continue executing in the background via `tokio::spawn`. This approach allows us to avoid the pitfalls of incomplete state transitions in critical futures without the overhead of spawning all futures immediately.

The `DropGuard` checks if the inner future has completed; if not, it safely spawns the future in the background during its drop. This solution strikes a balance between stack-based future management and the necessity to complete crucial futures.

Fixes https://github.com/TraceMachina/nativelink/issues/1156

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to
  not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

`cargo test --all` and `bazel test //...` passes as expected.

## Checklist

- [x] Updated documentation if needed
- [x] `bazel test //...`  passes locally
- [x] PR is contained in a single commit, using `git amend` see some [docs](https://www.atlassian.com/git/tutorials/rewriting-history)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/TraceMachina/nativelink/1300)
<!-- Reviewable:end -->

leodziki commented:
Hope to review the implementation of DropGuard soon
cc: @allada, @MarcusSorealheis

allada reviewed (changes requested):
> Reviewed 1 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/1300)* status: 0 of 2 LGTMs obtained, and 1 of 5 files reviewed, and 8 discussions need to be resolved

___
*[`nativelink-util/src/drop_guard.rs` line 1 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/1300#-O5gOqso3U22xfMeWDDA:-O5gOqso3U22xfMeWDDB:bok2db8) ([raw file](https://github.com/TraceMachina/nativelink/blob/f8c62b8ae79fc8a353f243ce4a7c0d1d0b591f57/nativelink-util/src/drop_guard.rs#L1)):*
> ```XML
> use std::future::Future;
> ```

nit: Needs copyright.
___
*[`nativelink-util/src/drop_guard.rs` line 5 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/1300#-O5gKOYvA_r_UQ97hJhs:-O5gKOYvA_r_UQ97hJht:b-njh44t) ([raw file](https://github.com/TraceMachina/nativelink/blob/f8c62b8ae79fc8a353f243ce4a7c0d1d0b591f57/nativelink-util/src/drop_guard.rs#L5)):*
> ```XML
> use std::task::{Context, Poll};
> 
> pub struct DropGuard<F: Future> {
> ```

nit: `F: Future + Send + 'static` because our future must be able to be transferred into a spawn
___
*[`nativelink-util/src/drop_guard.rs` line 5 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/1300#-O5gOeBb3A3HpJXzI9ZD:-O5gOeBb3A3HpJXzI9ZE:b-rzlv3j) ([raw file](https://github.com/TraceMachina/nativelink/blob/f8c62b8ae79fc8a353f243ce4a7c0d1d0b591f57/nativelink-util/src/drop_guard.rs#L5)):*
> ```XML
> use std::task::{Context, Poll};
> 
> pub struct DropGuard<F: Future> {
> ```

nit: Rename to `DropProtectedFuture` (and rename file).
___
*[`nativelink-util/src/drop_guard.rs` line 6 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/1300#-O5gK2Ad7NQAy0Nz9rJF:-O5gK2Ad7NQAy0Nz9rJG:b-4yu4ft) ([raw file](https://github.com/TraceMachina/nativelink/blob/f8c62b8ae79fc8a353f243ce4a7c0d1d0b591f57/nativelink-util/src/drop_guard.rs#L6)):*
> ```XML
> 
> pub struct DropGuard<F: Future> {
>     future: Option<Pin<Box<F>>>,
> ```

nit: We should be able to do this with a `F`. We should not need to `Box` and cannot `Pin` here (because we may xfer our future to a spawn).
___
*[`nativelink-util/src/drop_guard.rs` line 10 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/1300#-O5gNM2oEq0LrsrmTliF:-O5gNM2oEq0LrsrmTliG:bbi3kdh) ([raw file](https://github.com/TraceMachina/nativelink/blob/f8c62b8ae79fc8a353f243ce4a7c0d1d0b591f57/nativelink-util/src/drop_guard.rs#L10)):*
> ```XML
> 
> impl<F: Future> DropGuard<F> {
>     pub fn new(future: F) -> Self {
> ```

We need to be careful that we don't loose our global nativelink `Context` here.

1. We need to have the user pass in a span here.
2. The future that we store needs to be wrapped in`ContextAwareFuture::new` (plus the span the user passes in).
3. (see below in Drop call)
___
*[`nativelink-util/src/drop_guard.rs` line 30 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/1300#-O5gMzcU5kR86tIIfg3x:-O5gMzcU5kR86tIIfg3y:bh7augp) ([raw file](https://github.com/TraceMachina/nativelink/blob/f8c62b8ae79fc8a353f243ce4a7c0d1d0b591f57/nativelink-util/src/drop_guard.rs#L30)):*
> ```XML
>             }
>         } else {
>             panic!("Future already completed");
> ```

nit: use `unreachable`
___
*[`nativelink-util/src/drop_guard.rs` line 39 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/1300#-O5gN7EB-Zeo6j-Pv1BQ:-O5gN7EB-Zeo6j-Pv1BR:b1cnqvq) ([raw file](https://github.com/TraceMachina/nativelink/blob/f8c62b8ae79fc8a353f243ce4a7c0d1d0b591f57/nativelink-util/src/drop_guard.rs#L39)):*
> ```XML
>         if let Some(future) = self.future.take() {
>             // Block on the future to ensure it completes.
>             futures::executor::block_on(future);
> ```

This is unsafe, instead just forward it into a `spawn!()` and do not wait for the future.

Normally we'd require a `background_spawn!()`, but in this case we want to use `spawn!` directly and suppress the linter. If we use `background_spawn!` we'll create double work, because `ContextAwareFuture` handles all the context stuff we need to be careful with.
___
*[`nativelink-util/src/evicting_map.rs` line 438 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/1300#-O5gP6o_-HE8bjBsFdlH:-O5gOuhp8nvDwqhi35QT:bh544lf) ([raw file](https://github.com/TraceMachina/nativelink/blob/f8c62b8ae79fc8a353f243ce4a7c0d1d0b591f57/nativelink-util/src/evicting_map.rs#L438)):*
> ```XML
>             };
> 
>             let fut = state.put(key, eviction_item);
> ```

nit: We don't want to do this here, we want to make this entire future/function call protected. This can be done by rewriting the function signature to:
```rs
    fn inner_insert_many(
        &self,
        mut state: &mut State<K, T>,
        inserts: impl IntoIterator<Item = (K, T)>,
        seconds_since_anchor: i32,
    ) -> impl Future<Output=Vec<T>> + Send {
        DropSafeFuture::new(async move {
          // ...
        })
```


<!-- Sent from Reviewable.io -->

leodziki committed:
Implement `DropProtectedFuture` to ensure future completion

We introduced a `DropProtectedFuture` wrapper to manage state transitions
safely across `await` boundaries. It ensures that the future will continue
executing in the background via `tokio::spawn` even if a future is dropped
before completion.

leodziki reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/1300)* status: 0 of 2 LGTMs obtained, and 0 of 7 files reviewed, and pending CI: Bazel Dev / ubuntu-22.04, Cargo Dev / macos-13, Cargo Dev / ubuntu-22.04, Installation / macos-13, Installation / macos-14, Installation / ubuntu-22.04, Local / ubuntu-22.04, NativeLink.com Cloud / Remote Cache (Legacy Dockerfile Test), Publish image, Publish nativelink-worker-init, Publish nativelink-worker-lre-cc, Remote / large-ubuntu-22.04, asan / ubuntu-22.04, docker-compose-compiles-nativelink (22.04), integration-tests (22.04), macos-13, ubuntu-20.04 / stable, ubuntu-22.04, ubuntu-22.04 / stable, windows-2022 / stable, and 5 discussions need to be resolved (waiting on @allada)

___
*[`nativelink-util/src/evicting_map.rs` line 447 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/1300#-O6uHLSW4tzPNdLqTTmO:-O6uHLSW4tzPNdLqTTmP:b8ljxq) ([raw file](https://github.com/TraceMachina/nativelink/blob/fcc7fd522f4801c11405c2d5a33f50a1e8d511a3/nativelink-util/src/evicting_map.rs#L447)):*
> ```XML
>                     };
> 
>                     if let Some(old_item) = state.put(key, eviction_item).await {
> ```

I'm submitting this WIP to clarify unclear points.

As we see, there is `self` and `state` referenced here & mutated n this `async move {}` block.

We are gonna implement `Send + `static` trait on the Future itself.

How shall we handle this?

Should we change this functions parameter as `Arc<Mutex>` for state and add `Arc<Self>` as an additional parameter for `self`?
Or should we implement `Clone` trait on `self` and `state`?

Waiting for clarifying further walks from this.

Thanks.
___
*[`nativelink-util/src/evicting_map.rs` line 450 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/1300#-O6uHi6mDPz9xeGx2ELz:-O6uHi6n-aneAwVw4nWA:b-875k1z) ([raw file](https://github.com/TraceMachina/nativelink/blob/fcc7fd522f4801c11405c2d5a33f50a1e8d511a3/nativelink-util/src/evicting_map.rs#L450)):*
> ```XML
>                         replaced_items.push(old_item);
>                     }
>                     state.sum_store_size += new_item_size;
> ```

Here.
___
*[`nativelink-util/src/evicting_map.rs` line 452 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/1300#-O6uHdlo5DjZX27pAtpO:-O6uHdlo5DjZX27pAtpP:b-875k1z) ([raw file](https://github.com/TraceMachina/nativelink/blob/fcc7fd522f4801c11405c2d5a33f50a1e8d511a3/nativelink-util/src/evicting_map.rs#L452)):*
> ```XML
>                     state.sum_store_size += new_item_size;
>                     state.lifetime_inserted_bytes.add(new_item_size);
>                     self.evict_items(state.deref_mut()).await;
> ```

Here.
___
*[`nativelink-util/src/drop_guard.rs` line 10 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/1300#-O5gNM2oEq0LrsrmTliF:-O6uH-DcCKa3TvMHCmXP:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/f8c62b8ae79fc8a353f243ce4a7c0d1d0b591f57/nativelink-util/src/drop_guard.rs#L10)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

We need to be careful that we don't loose our global nativelink `Context` here.

1. We need to have the user pass in a span here.
2. The future that we store needs to be wrapped in`ContextAwareFuture::new` (plus the span the user passes in).
3. (see below in Drop call)
</blockquote></details>

Done.
___
*[`nativelink-util/src/drop_guard.rs` line 39 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/1300#-O5gN7EB-Zeo6j-Pv1BQ:-O6uGzr-96bsN9ZtEton:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/f8c62b8ae79fc8a353f243ce4a7c0d1d0b591f57/nativelink-util/src/drop_guard.rs#L39)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

This is unsafe, instead just forward it into a `spawn!()` and do not wait for the future.

Normally we'd require a `background_spawn!()`, but in this case we want to use `spawn!` directly and suppress the linter. If we use `background_spawn!` we'll create double work, because `ContextAwareFuture` handles all the context stuff we need to be careful with.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

leodziki commented:
Waiting for opinion, @allada
